docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,184 @@
+# Memory allocation routines
+## What is the problem?
+FEX-Emu needs to allocate memory differently than regular applications. This problem happens because FEX runs both 32-bit and 64-bit guest
+applications in the same address space as FEX itself. When running 32-bit applications, FEX reserves up all memory above 4GB in order to correctly
+emulate the 32-bit address space. We then use that reserved space for FEX allocations, so we don't interrupt application's own allocations.
+
+### Why not just just replace the system allocator?
+jemalloc and similar allocators allow your application to replace the system allocator. This is something that FEX used to do, but this has since been
+removed because 32-bit thunks need to allocate memory in the lower 4GB.
+Overriding the system allocator is a global operation, including any libraries get that opened at runtime. This would cause thunks to use FEX's
+allocators. This would interfere with 32-bit thunks in such a way that their memory allocations would never end up in 32-bit virtual address space.
+Allocations made in native host libraries must always be valid 32-bit pointers for these 32-bit thunks to operate.
+
+With this approach, we couldn't find an easy way to customize allocation behaviour for the requirements of 32-bit thunking.
```

Comment:
This is a bit misleading now since we do use jemalloc to override the system allocator again (probably best to just drop the historical note).

I think I found a good concise wording that also streamlines this a little bit in general to first mention the requirements of 32-bit thunks and then the subsequent problem with overriding the allocator:

```suggestion
We could control the placement of FEX's internal allocations by overriding the system
allocator. However, 32-bit thunks (and their corresponding native host libraries) still
need to allocate memory in the lower 4 GB of memory so that they produce
guest-accessible pointers. Since overriding the system allocator is a global operation,
selectively overriding it like this is not possible.

Since we found no way to resolve this conflict, we had to resort to the alternative of avoiding use of the system allocator for FEX's internal allocations entirely (where possible).
```

Does that sound reasonable?

Reply:
Sounds reasonable to me. Switched over to it.